### PR TITLE
library: Add Popover entry

### PR DESCRIPTION
**Pull request by Nasah-Kuma**

### CODE REVIEW

File: src/Library/demos/Popovers/main.blp [hunk 1]
```diff
@@ -0,0 +1,91 @@
+using Gtk 4.0;
+using Adw 1;
+
+Adw.StatusPage {
+  title: "Popovers";
+  description: _("Display content in a container above all other content on the window");
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.blp:6]:
```suggestion
  description: _("Display content in a container anchored to an other widget");
```

File: src/Library/demos/Popovers/main.blp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    Box {
+      margin-bottom: 80;
+      margin-top: 40;
+
+      MenuButton popover_button {
+      name: 'popover_button';
+      popover: plain_popover;
+      label: "plain popover";
+      primary: true;
+      margin-end: 40;
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.blp:21]:
```suggestion
        name: 'popover_button';
        popover: plain_popover;
        label: "Plain Popover";
        primary: true;
        margin-end: 40;
```

File: src/Library/demos/Popovers/main.blp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      }
+
+      MenuButton popover_menu_button {
+        name: "popover_menu_button";
+        label: "Popover menu";
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.blp:26]: please use Title Case for labels

```suggestion
        label: "Popover Menu";
```

File: src/Library/demos/Popovers/main.blp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    LinkButton {
+      label: "Popover menu";
+      uri: "https://docs.gtk.org/gtk4/class.PopoverMenu.html";
+    }
+
+    LinkButton {
+      label: "Popover menu bar";
+      uri: "https://docs.gtk.org/gtk4/class.PopoverMenuBar.html";
+    }
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.blp:50]: please remove this one - we don't want/use menu bars in GNOME

File: src/Library/demos/Popovers/main.blp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      styles ["title-4"]
+    }
+
+    LinkButton {
+      label: "Popover";
+      uri: "https://docs.gtk.org/gtk4/class.Popover.html";
+    }
+
+    LinkButton {
+      label: "Popover menu";
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.blp:43]:
```suggestion
      label: "Popover Menu";
```

File: src/Library/demos/Popovers/main.blp [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        label: "Popover menu";
+        margin-end: 40;
+        popover: popover_menu;
+      }
+    }
+
+    Label sub_heading {
+      label: "API References";
+      styles ["title-4"]
+    }
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.blp:35]: I think we can remove this

File: src/Library/demos/Popovers/main.blp [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+      uri: "https://docs.gtk.org/gtk4/class.PopoverMenu.html";
+    }
+
+    LinkButton {
+      label: "Popover menu bar";
+      uri: "https://docs.gtk.org/gtk4/class.PopoverMenuBar.html";
+    }
+
+    LinkButton {
+      margin-top: 20;
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.blp:53]: please use only multiples of 6

it's a GNOME design guideline

 - `24` works well here

File: src/Library/demos/Popovers/main.blp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+Popover plain_popover {
+  has-arrow : true;
+  name: 'Plain popover';
+  child: Box plain-popover-box {
+    name: 'plain-popover-box';
+    Label{
+      label: 'Plain popover';
+          }
+        };
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.blp:68]: please fix indentation

File: src/Library/demos/Popovers/main.blp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  }
+}
+
+Popover plain_popover {
+  has-arrow : true;
+  name: 'Plain popover';
+  child: Box plain-popover-box {
+    name: 'plain-popover-box';
+    Label{
+      label: 'Plain popover';
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.blp:66]:
```suggestion
      label: 'Plain Popover';
```

File: src/Library/demos/Popovers/main.blp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+      margin-top: 20;
+      label: "Human Interface Guidelines";
+      uri: "https://developer.gnome.org/hig/patterns/containers/popovers.html";
+    }
+  }
+}
+
+Popover plain_popover {
+  has-arrow : true;
+  name: 'Plain popover';
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.blp:62]: this is not a string for humans

File: src/Library/demos/Popovers/main.blp [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+          }
+        };
+}
+
+PopoverMenu popover_menu{
+  name: 'Popover menu';
+  menu-model: menu_app;
+}
+
+menu menu menu_app {
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.blp:76]:
```suggestion
menu menu_app {
```

File: src/Library/demos/Popovers/main.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+const popover_ids = [
+"plain_popover",
+"popover_menu"
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.js:3]: if you press Run it will format the JavaScript

File: src/Library/demos/Popovers/main.json [hunk 13]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "Popovers",
+  "category": "user_interface",
+  "description": "Popups are used to display content in a container above all other content on the window",
+  "panels": [
+    "code",
+    "preview"
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.json:7]: `ui` is more important here

```suggestion
    "ui",
    "preview"
```

File: src/Library/demos/Popovers/main.json [hunk 14]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "Popovers",
+  "category": "user_interface",
+  "description": "Popups are used to display content in a container above all other content on the window",
```
Comment (sonnyp) [inline src/Library/demos/Popovers/main.json:4]: same as above :pray:

**Review by sonnyp — COMMENTED** [review]
Thanks! Great start.

Could you please also center the buttons relative to the title? I think you can simplify the layout too by using `spacing` in the buttons parent container. 

image

File: src/Library/demos/Popovers/main.blp (hunk 1 above)
Comment (Nasah-Kuma) [reply src/Library/demos/Popovers/main.blp:6]: Display content in a container anchored to another widget, right?

Comment (Nasah-Kuma) [reply src/Library/demos/Popovers/main.blp:6]: Display content in a container anchored to another widget, right?

File: src/Library/demos/Popovers/main.blp (hunk 10 above)
Comment (Nasah-Kuma) [reply src/Library/demos/Popovers/main.blp:62]: Not sure I understand what you mean

File: src/Library/demos/Popovers/main.blp (hunk 1 above)
Comment (sonnyp) [reply src/Library/demos/Popovers/main.blp:6]: Yes – did I make a mistake in my suggestion? I can't spot it.

File: src/Library/demos/Popovers/main.blp (hunk 10 above)
Comment (sonnyp) [reply src/Library/demos/Popovers/main.blp:62]: `name` should be an identifier, like an id or the name of a variable

File: src/Library/demos/Popovers/main.blp (hunk 8 above)
Comment (sonnyp) [reply src/Library/demos/Popovers/main.blp:68]: Unfortunately, we don't have a formatter for blueprint yet

trick: you can convert to XML and back to Blueprint

Comment (Nasah-Kuma) [reply src/Library/demos/Popovers/main.blp:68]: ok thanks

File: src/Library/demos/Popovers/main.blp (hunk 10 above)
Comment (Nasah-Kuma) [reply src/Library/demos/Popovers/main.blp:62]: Like this plain_popover?

Comment (sonnyp) [reply src/Library/demos/Popovers/main.blp]: yes, but please be consistent – either `-` or `_` for identifiers

File: src/Library/demos/Popovers/main.blp (hunk 7 above)
Comment (sonnyp) [reply src/Library/demos/Popovers/main.blp]: there are still spacings that are not multiples of 6

Comment (Nasah-Kuma) [reply src/Library/demos/Popovers/main.blp]: Oh sorry! I missed this

File: src/Library/demos/Popovers/main.js [hunk 15]
```diff
@@ -1,14 +1,15 @@
...
 for (const id of popover_ids) {
   const popover = workbench.builder.get_object(id);
   popover.connect("closed", onClosed);
 }
 
 function onClosed(popover) {
-  console.log(`${popover.name} closed.`);
+  console.log(
+    `${
+      popover.name === "plain_popover" ? "plain popover" : "popover menu"
```
[the remark this answers is not in the document]
Comment (Nasah-Kuma) [reply src/Library/demos/Popovers/main.js:11]: I thought plain over looks better on the console than plain_popover but I'll have it changed.

File: src/Library/demos/Popovers/main.blp (hunk 11 above)
Comment (sonnyp) [reply src/Library/demos/Popovers/main.blp]: This is not resolved and prevents Workbench from building

> error: Unexpected tokens
at ../src/./Library/demos/Popovers/main.blp line 76 column 1:
  76 |menu menu menu_app {

Please make sure you resolved all previous comments and re-request review when it's ready

image

Additionally, please make sure that you test your changes before submitting for review.

**Review by sonnyp — CHANGES_REQUESTED** [review]
https://github.com/sonnyp/Workbench/pull/249#discussion_r1155167364

**Review by sonnyp — APPROVED** [review]
Thanks!

I took the liberty of adding you in the list of contributors. https://github.com/sonnyp/Workbench/pull/249/commits/4873a605be3d856963de8af7de2b758f6f06628e

**Comment by Nasah-Kuma** [comment]
> Thanks!
> 
> I took the liberty of adding you in the list of contributors. [4873a60](https://github.com/sonnyp/Workbench/commit/4873a605be3d856963de8af7de2b758f6f06628e)

Great!

**Pull request merged**